ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java

```diff
@@ -137,17 +137,69 @@ private void prepareReturnValues(List<String> values) throws SemanticException {
     Utils.writeOutput(values, new Path(work.resultTempPath), conf);
   }
 
+  /**
+   * Decide whether to examine all the tables to dump. We do this if
+   * 1. External tables are going to be part of the dump : In which case we need to list their
+   * locations.
+   * 2. External or ACID tables are being bootstrapped for the first time : so that we can dump
+   * those tables as a whole.
+   * @return
+   */
+  private boolean shouldDumpExternalTableLocation() {
+    return conf.getBoolVar(HiveConf.ConfVars.REPL_INCLUDE_EXTERNAL_TABLES)
+            && (!conf.getBoolVar(HiveConf.ConfVars.REPL_DUMP_METADATA_ONLY)
+            || conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_EXTERNAL_TABLES));
+  }
+
+  private boolean shouldExamineTablesToDump() {
+    return shouldDumpExternalTableLocation() ||
+            conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_ACID_TABLES);
+  }
+
+  private boolean shouldBootstrapDumpTable(Table table) {
+    if (conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_EXTERNAL_TABLES) &&
+            TableType.EXTERNAL_TABLE.equals(table.getTableType())) {
+      return true;
+    }
+
+    if (conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_ACID_TABLES) &&
+           AcidUtils.isTransactionalTable(table)) {
+      return true;
+    }
+
+    return false;
+  }
+
   private Long incrementalDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive hiveDb) throws Exception {
     Long lastReplId;// get list of events matching dbPattern & tblPattern
     // go through each event, and dump out each event to a event-level dump dir inside dumproot
+    ValidTxnList validTxnList = null;
+    long waitUntilTime = 0;
+    long bootDumpBeginReplId = -1;
+
+    // If we are bootstrapping ACID tables, we need to perform steps similar to a regular
+    // bootstrap (See bootstrapDump() for more details. Only difference here is instead of
+    // waiting for the concurrent transactions to finish, we start dumping the incremental events
+    // and wait only for the remaining time if any.
+    if (conf.getBoolVar(HiveConf.ConfVars.REPL_BOOTSTRAP_ACID_TABLES)) {
+      bootDumpBeginReplId = queryState.getConf().getLong(ReplUtils.LAST_REPL_ID_KEY, -1L);
+      assert (bootDumpBeginReplId >= 0);
+      LOG.info("Dump for bootstrapping ACID tables during an incremental dump for db {} and table {}",
+              work.dbNameOrPattern,
+              work.tableNameOrPattern);
+      validTxnList = getTxnMgr().getValidTxns();
```

Comment:
validTxnList can be obtained inside getValidTxnListForReplDump(). No need to get it by caller and pass it. It is enough to check the status of open txns at that time. It will also avoid an additional iteration if many txns are closed during incremental events dump. Only need to get and pass waitUntilTime.

Reply:
If we take the snapshot after dumping events, the transactions opened while dumping events won't get enough time to complete since the waiting time is reduced by the time taken for dumping events. By taking the transaction snapshot before dumping events, all the transactions opened then get configured time to complete. That's why we pass both the snapshot and waiting time reduced by the event dump time to getValidTxnListForReplDump().

Furthermore this will widen the time gap between noting down the last repl id and taking the snapshot, thus increasing the chances of hitting the bug we discussed in earlier round of review comments.

Reply:
As discussed, getting ValidTxnList inside getValidTxnListForReplDump() doesn't make any difference as it doesn't include the txns opened after currentTxn. Refer to TxnCommonUtils.createValidReadTxnList() for this.
So, please revert back the changes as proposed in my comment.

Reply:
Thanks for the explanation. Done.